sklearn/ensemble/forest.py

```diff
@@ -402,12 +418,66 @@ def __init__(self,
             estimator_params=estimator_params,
             bootstrap=bootstrap,
             oob_score=oob_score,
+            permutation_feature_importances=permutation_feature_importances,
             n_jobs=n_jobs,
             random_state=random_state,
             verbose=verbose,
             warm_start=warm_start,
             class_weight=class_weight)
 
+    def _set_permutation_feature_importances(self, X, y):
+        # Convert data to arrays, y to row vector
+        X = np.array(X)
+        y = np.squeeze(np.array(y))
+
+        if y.ndim != 1:
+            raise NotImplementedError('permutation feature importances not '
+                                      'implemented for multiple output problems.')
+
+        n_features = X.shape[1]
+        random_state = check_random_state(self.random_state)
+
+        forest = self.estimators_
+
+        for tree in forest:
+            X_oob, y_oob, n_oob = self._get_oob_data(X, y, tree)
+            y_oob_pred = self._predict_oob(X_oob, tree)
+
+            feature_importances = np.zeros(n_features)
+            for feature_ind in xrange(n_features):
+                y_oob_pred_perm = self._predict_oob(X_oob, tree,
+                                                    shuffle_ind=feature_ind,
+                                                    random_state=random_state)
+                n_wrong = self._calc_mislabel_rate(y_oob, y_oob_pred)
+                n_wrong_perm = self._calc_mislabel_rate(y_oob, y_oob_pred_perm)
+
+                feature_importances[feature_ind] = (n_wrong_perm - n_wrong) / n_oob
+
+            tree._feature_importances = feature_importances
+
+    @staticmethod
+    def _get_oob_data(X, y, tree):
+        n_samples = X.shape[0]
+        oob_ind = _generate_unsampled_indices(tree.random_state, n_samples)
+
+        n_oob = len(oob_ind)
+        X_oob = X[oob_ind, :]
+        y_oob = y[oob_ind]
+        return X_oob, y_oob, n_oob
+
+    @staticmethod
+    def _predict_oob(feature_data, tree, shuffle_ind=None, random_state=None):
+        features_tmp = feature_data.copy()
+        if shuffle_ind is not None:
+            features_tmp[:, shuffle_ind] = shuffle(features_tmp[:, shuffle_ind],
+                                                   random_state=random_state)
+        return tree.predict(features_tmp)
+
+    @staticmethod
+    def _calc_mislabel_rate(expected, predicted):
+        bool_mislabel = np.squeeze(expected) != np.squeeze(predicted)
```

Comment:
in what case should we expect `expected.shape != predicted.shape` that we should want to squeeze?